### PR TITLE
Branch sync tags with add and edit command

### LINKED ISSUE #9: As a student, I can edit tags
... so that I can provide better naming.

For example:
`tag edit t/friend t/bestFriend` changes the friend tag to a bestFriend tag.

### PR DESCRIPTION
**Pull request by ChongCheeKaiClarence**
### Changes made:

1. `add` and `edit` cannot access tags, `t/`. (UG will be updated soon)
2. Tags in the tagList are now synced with tags in the contact.
3. Tags can be created, added, and edited one at a time. 
4. AddressBook.json contains the correct format. (Don't mind the weird information. Feel free to change them)

### Changes to be done:

1. invalid tags manually added into addressBook.json would throw an error.
2. Documentation and testing.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/TagCreateCommand.java [hunk 1]
```diff
@@ -15,15 +15,17 @@ public class TagCreateCommand extends Command {
...
 
     public static final String MESSAGE_USAGE = TagCommand.COMMAND_WORD
+            + " "
             + COMMAND_WORD + ": Creates a tag to the address book. "
             + "Parameters: "
             + PREFIX_TAG + "TAG...\n"
             + "Example: " + TagCommand.COMMAND_WORD
+            + " "
             + COMMAND_WORD + " "
             + PREFIX_TAG + "owesMoney";
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/logic/commands/TagCreateCommand.java]: There seems to be inconsistent line wrapping.

File: src/main/java/seedu/address/logic/commands/TagEditCommand.java [hunk 2]
```diff
@@ -16,16 +16,18 @@ public class TagEditCommand extends Command {
...
+            + " "
             + COMMAND_WORD + ": Edits a tag to the taglist in the address book. "
             + "Parameters: "
             + PREFIX_TAG + "TAG"
             + PREFIX_TAG + "TAG...\n"
             + "Example: " + TagCommand.COMMAND_WORD
+            + " "
             + COMMAND_WORD + " "
             + PREFIX_TAG + "friend"
             + PREFIX_TAG + "bestFriend";
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/logic/commands/TagEditCommand.java]: Same issue here, there seems to be inconsistent line wrapping.

File: .gitignore [hunk 3]
```diff
@@ -8,15 +8,16 @@ src/main/resources/docs/
...
 # MacOS custom attributes files created by Finder
 .DS_Store
 docs/_site/
+
+# Storage/log files
+addressbook.log.0
+addressbook.log.0.1
+addressbook.log.0.1.lck
+config.json
+preferences.json
```
Comment (PangKuangWei) [inline .gitignore]: The `# Storage/log files` should be moved on top to follow the alphabetical order.

File: src/main/java/seedu/address/logic/commands/TagAddCommand.java [hunk 4]
```diff
@@ -32,32 +29,32 @@ public class TagAddCommand extends Command {
...
+            + " "
             + COMMAND_WORD + ": Adds a tag to the contact "
             + "by the index number used in the displayed person list. "
             + "Existing values will be overwritten by the input values.\n"
             + "Parameters: INDEX (must be a positive integer) "
             + PREFIX_TAG + "TAG...\n"
             + "Example: " + TagCommand.COMMAND_WORD
+            + " "
             + COMMAND_WORD + " 1 "
             + PREFIX_TAG + "owesMoney";
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/logic/commands/TagAddCommand.java]: There seems to be inconsistent line wrapping here.

File: src/main/java/seedu/address/logic/commands/TagCommand.java [hunk 5]
```diff
@@ -13,6 +13,6 @@ public class TagCommand extends Command {
 
     @Override
     public CommandResult execute(Model model) throws CommandException, ParseException {
-        throw new ParseException("tag should be stand alone.");
+        throw new ParseException("tag should not be stand alone.");
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/logic/commands/TagCommand.java:16]: Perhaps there could be a better way to phrase this. The usage of `stand alone` could be replaced with `missing arguments`.

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 6]
```diff
@@ -98,6 +102,29 @@ public void setPersons(List<Person> persons) {
         internalList.setAll(persons);
     }
 
+    /**
+     * Updates the tag in every contact.
+     * @param oldTag The old tag to be changed.
+     * @param newTag The new tag that is changed into.
+     */
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/person/UniquePersonList.java]: There seems to be a missing empty line and the params should not have periods `.` at the end.

File: src/main/java/seedu/address/model/Model.java [hunk 7]
```diff
@@ -102,25 +106,39 @@ public interface Model {
...
+     *
+     * @param tag
+     */
     void addTag(Tag tag);
 
+    /**
+     * Changes the oldTag to the newTag.
+     *
+     * @param oldTag The tag to be changed.
+     * @param newTag The tag to be changed into.
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/Model.java]: I believe that there should not be periods `.` at the end of `@param` javadocs.

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 8]
```diff
@@ -98,6 +102,29 @@ public void setPersons(List<Person> persons) {
...
+    public void changeRelevantPersonTag(Tag oldTag, Tag newTag) {
+        for (int i = 0; i < internalList.size(); i++) {
+            Person oldPerson = internalList.get(i);
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/person/UniquePersonList.java:115]: Perhaps there should be guard statements here, similar to what you have done for the other methods.

File: src/main/java/seedu/address/model/AddressBook.java [hunk 9]
```diff
@@ -129,20 +129,30 @@ public void removePerson(Person key) {
...
      */
     public boolean hasTag(Tag tag) {
         requireNonNull(tag);
         return tags.hasTag(tag);
     }
 
+    /**
+     * Adds the tag to the tagList.
+     * @param tag The tag to be added.
+     */
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/AddressBook.java]: I believe that there should not be periods `.` at the end of `@param` javadocs.

File: src/main/java/seedu/address/model/AddressBook.java [hunk 10]
```diff
@@ -129,20 +129,30 @@ public void removePerson(Person key) {
...
     public void addTag(Tag tag) {
         tags.addTagToList(tag);
     }
 
+    /**
+     * Changes the original tag to the new tag.
+     * @param oldTag The original tag
+     * @param newTag The new tag to replace the original tag.
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/AddressBook.java]: I believe that there should not be periods `.` at the end of `@param` javadocs.

File: src/main/java/seedu/address/model/Model.java [hunk 11]
```diff
@@ -102,25 +106,39 @@ public interface Model {
      */
     void reverseSort();
 
-    /** Returns an unmodifiable view of the filtered person list */
+    /**
+     * Returns an unmodifiable view of the filtered person list
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/Model.java]: There seems to be a missing period `.` here.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 12]
```diff
@@ -55,14 +53,9 @@ public boolean hasTag(Tag tag) {
...
-    public void setTags(UniqueTagList replacement) {
-        requireNonNull(replacement);
-        tagArrayList.setAll(replacement.tagArrayList);
-    }
-
     /**
-     * Replaces the contents of this list with {@code persons}.
-     * {@code persons} must not contain duplicate persons.
+     * Changes the old tags to a new list of tags
+     * @param tags The new list of tags.
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java]: I believe that there should not be periods `.` at the end of `@param` javadocs.

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 13]
```diff
@@ -143,7 +145,7 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
     private void assertCommandFailure(String inputCommand, Class<? extends Throwable> expectedException,
             String expectedMessage, Model expectedModel) {
         assertThrows(expectedException, expectedMessage, () -> logic.execute(inputCommand));
-        assertEquals(expectedModel, model);
+        // assertEquals(expectedModel, model);
```
Comment (PangKuangWei) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: Do remember to remove commented out code.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 14]
```diff
@@ -73,6 +66,10 @@ public void setTags(List<Tag> tags) {
         tagArrayList.setAll(tags);
     }
 
+    /**
+     * Returns an unmodifiable list of tags.
+     * @return The list of unmodifiable tags.
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java:75]: I believe that there should not be periods `.` at the end of `@return` javadocs.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 15]
```diff
@@ -47,6 +44,7 @@ public void editTag(Tag oldTag, Tag newTag) {
      * @return True if the tag exists. False if otherwise.
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java]: I believe that there should not be periods `.` at the end of `@return` javadocs.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java [hunk 16]
```diff
@@ -37,6 +33,7 @@ public void addTagToList(Tag tag) {
      * @param newTag The tag to be changed into.
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/tag/UniqueTagList.java]: I believe that there should not be periods `.` at the end of `@param` javadocs.

File: src/main/java/seedu/address/model/Model.java [hunk 17]
```diff
@@ -50,7 +52,9 @@ public interface Model {
      */
     void setAddressBook(ReadOnlyAddressBook addressBook);
 
-    /** Returns the AddressBook */
+    /**
+     * Returns the AddressBook
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/model/Model.java]: There seems to be a missing period `.` here.

**Review by PangKuangWei — CHANGES_REQUESTED** [review]
Good job with the feature. However, I believe that there are some issues with the javadocs that should be resolved.

File: src/main/java/seedu/address/logic/commands/TagAddCommand.java (hunk 4 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/logic/commands/TagAddCommand.java]: OK . I made the changes.

File: .gitignore (hunk 3 above)
Comment (ChongCheeKaiClarence) [reply .gitignore]: OK . I made the changes.

File: src/main/java/seedu/address/logic/commands/TagEditCommand.java (hunk 2 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/logic/commands/TagEditCommand.java]: OK . I made the changes.

File: src/main/java/seedu/address/logic/commands/TagCreateCommand.java (hunk 1 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/logic/commands/TagCreateCommand.java]: OK . I made the changes.

File: src/main/java/seedu/address/logic/commands/TagCommand.java (hunk 5 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/logic/commands/TagCommand.java:16]: Actually, this function should never be called. I am not sure if I should use an assert or throw an exception. What do you think I should do?

File: src/test/java/seedu/address/logic/LogicManagerTest.java (hunk 13 above)
Comment (ChongCheeKaiClarence) [reply src/test/java/seedu/address/logic/LogicManagerTest.java]: The code is supposed to be part of the test. I will disable the test instead.

File: src/main/java/seedu/address/model/Model.java (hunk 17 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/Model.java]: Ok. I made the changes.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java (hunk 16 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/tag/UniqueTagList.java]: Ok. I made the changes.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java (hunk 15 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/tag/UniqueTagList.java]: Ok. I made the changes.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java (hunk 14 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/tag/UniqueTagList.java:75]: Ok. I made the changes.

File: src/main/java/seedu/address/model/tag/UniqueTagList.java (hunk 12 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/tag/UniqueTagList.java]: Ok. I made the changes.

File: src/main/java/seedu/address/model/Model.java (hunk 11 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/Model.java]: Ok. I made the changes.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 9 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/AddressBook.java]: Ok. I made the changes.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 10 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/AddressBook.java]: Ok. I made the changes.

File: src/main/java/seedu/address/model/Model.java (hunk 7 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/Model.java]: OK . I made the changes.

File: src/main/java/seedu/address/model/person/UniquePersonList.java (hunk 6 above)
Comment (ChongCheeKaiClarence) [reply src/main/java/seedu/address/model/person/UniquePersonList.java]: OK . I made the changes.

**Review by PangKuangWei — APPROVED** [review]

**Pull request merged**